Subscriptions: switch to sliding buffer, avoid too many puts, last value wins

oliyh opened the pull request:
Hi,

I'd like to get your opinion on this change. Background is that some queries can be very expensive, and the streamer may invoke the `source-stream` function faster than the query can be completed. This results in a build-up of pending puts on this channel which can eventually exceed 1024 and an exception is thrown.

Throttling on the streamer is of course possible but it is blind - you can debounce the `source-stream` function but you still can't know how long the query took to execute, and the function may easily be invoked again before the previous query has been completed.

This change prevents the pending puts from building up by adopting a last-one-wins attitude. It assumes that the client is not going to be interested in intermediate responses when there are newer ones just about to be delivered. This seems reasonable to me but you may think differently.

Further work could allow the size of the sliding buffer to be configured, or the addition of a synchronous version of `source-stream` which would only complete when the query had been completed, or some other mechanism to feed back the completion of the query to the streamer.

I haven't looked too closely at how the change introduced here could be explicitly tested, but it does pass all the existing tests.

Thanks

hlship commented:
I'll look into this.  I suspect this is going to be a place where we need to punt a bit, and make it a configurable option somehow ... one size likely does not fit all!

oliyh commented:
This second change actually goes hand in hand with the first. With these two changes:

- A race condition, which could arise when a subscription firing very quickly launches multiple async threads to perform the queries which complete in a non-deterministic order, is avoided
- Unbounded growth of async threads is avoided, and sized to one thread per subscription
- Queued interim values which will be immediately superceded are dropped, helping to catch up with the latest value more quickly and reducing compute resources needed for the subscription

Hope this makes sense, please let me know if you need any clarification.

Cheers

oliyh commented:
Hi @hlship 
I have added some documentation. I squashed the commit which has possibly confused Github a bit but you can see the changes here:  https://github.com/walmartlabs/lacinia-pedestal/pull/56/files?short_path=1066c61#diff-1066c615a05cba0a206d7127848fc139

Hope it's sufficient, let me know if you want me to clarify anything further.

hlship commented:
This has been replaced by #68
